rpk: add `registry` command to support the schema registry

r-vasquez opened the pull request:
WIP

This works entirely, but could do with some tests and potentially some thought on output.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md##pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

emaxerrno commented:
this is cool.

r-vasquez commented on src/go/rpk/pkg/cli/registry/schema/check_compatibility.go:
```diff
@@ -0,0 +1,80 @@
+				out.MaybeDie(err, "unable to print in the required format %q: %v", f.Kind, err)
+				out.Exit(s)
+			}
+			if compatible {
+				out.Exit("Schema is compatible.")
+			} else {
+				out.Die("Schema is not compatible.")
+			}
```
I guess we should always exit(0) if we successfully checked compatibility.

twmb commented on src/go/rpk/pkg/config/params.go:
```diff
@@ -912,7 +1001,7 @@ func (*Params) backcompatOldCloudYaml(fs afero.Fs) error {
+		} else if rpkYaml.Version > 2 {
 			return fmt.Errorf("%s is using a newer rpk.yaml format than we understand, please upgrade rpk", def)
 		}
 	}
```
I think after this, we need to set rpkYaml.Version to 2 always

twmb commented on src/go/rpk/pkg/config/params.go:
```diff
@@ -985,7 +1074,7 @@ func (p *Params) readRpkConfig(fs afero.Fs, c *Config) error {
 		c.rpkYaml = before // this config is not an rpk.yaml; preserve our defaults
 		return nil
-	} else if c.rpkYaml.Version > 1 {
+	} else if c.rpkYaml.Version > 2 {
```
I think after this, we need to set rpkYaml.Version to 2 always

twmb commented on src/go/rpk/pkg/cli/registry/schema/check_compatibility.go:
```diff
@@ -0,0 +1,86 @@
+		},
+	}
+
+	cmd.Flags().StringVar(&schemaFile, "schema", "", "Schema filepath to upload, must be .avro or .proto")
```
s/upload/check/ ?

twmb commented on src/go/rpk/pkg/cli/registry/schema/create.go:
```diff
@@ -0,0 +1,77 @@
+
+This uploads a schema to the registry, creating the schema if it does not
+exist. The schema type is detected by the filename extension: ".avro" for Avro
+and ".proto" for Protobuf. You can manually specify the type with the --type
```
The --type flag does not exist?

r-vasquez replied:
I'm going to delete this part of the doc, or do we want a `type` flag and avoid the file extension as a hard requirement?

twmb replied:
--type might actually be valuable so we don't require specific file extensions

twmb commented on src/go/rpk/pkg/cli/registry/schema/check_compatibility.go:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	cmd.Flags().StringVar(&schemaFile, "schema", "", "Schema filepath to upload, must be .avro or .proto")
+	cmd.Flags().StringVar(&sversion, "version", "", "Version to check compatibility with")
```
For this flag and all version flags, can we add `(latest, 0, 1...)` to the help text?

Also, doesn't this --version flag conflict with the global --version flag on rpk? This might need to be schema-version? (same comment for all commands)

twmb commented on src/go/rpk/pkg/cli/registry/schema/delete.go:
```diff
@@ -0,0 +1,65 @@
+
+func newDeleteCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+	var sversion string
+	var permanent bool
```
Can this be `isPermanent`? It better reads as a bool in the usage below

twmb commented on src/go/rpk/pkg/cli/registry/schema/delete.go:
```diff
@@ -0,0 +1,65 @@
+		},
+	}
+
+	cmd.Flags().StringVar(&sversion, "version", "", "Version to delete, -1 or latest deletes the latest version")
```
This command has a bit more help text in the version to delete info, so this might be better than my suggestion of `(latest, 0, 1...)` (and this wording also helpfully indicates that -1 deletes latest)

r-vasquez replied:
Actually I prefer not mentioning `-1` for the POSIX issue we raised before so instead I went with:

```Schema version to delete (latest, 0, 1...)```

twmb commented on src/go/rpk/pkg/cli/registry/subject.go:
```diff
@@ -0,0 +1,144 @@
+				out.MaybeDie(err, "unable to print in the required format %q: %v", f.Kind, err)
+				out.Exit(s)
+			}
+			for _, s := range subjects {
```
sort the subjects first?

twmb commented on src/go/rpk/pkg/cli/registry/subject.go:
```diff
@@ -0,0 +1,144 @@
+	return cmd
+}
+
+type deleteResponse struct {
```
For what it's worth, the types that are only used within a single function can be defined within the function itself.

r-vasquez replied:
Yes, I just don't want to stuff everything at the beginning of the command `Run` function and I tried to leave "only the logic" we would end up with:

```
		Run: func(cmd *cobra.Command, subjects []string) {
			type schemaResponse struct {
				Subject string `json:"subject" yaml:"subject"`
				Version int    `json:"version,omitempty" yaml:"version,omitempty"`
				ID      int    `json:"id,omitempty" yaml:"ID,omitempty"`
				Type    string `json:"type,omitempty" yaml:"type,omitempty"`
				Err     string `json:"error,omitempty" yaml:"error,omitempty"`
			}
			f := p.Formatter
			if h, ok := f.Help([]schemaResponse{}); ok {
				out.Exit(h)
			}
```

Before even starting with the logic.

I know it can be inside of the `newListCommand` function but if it's ok for you, I'll leave it outside.

twmb replied:
Putting it inside newDeleteCommand is what I meant, but I'm not too attached.

twmb commented on src/go/rpk/pkg/cli/registry/subject.go:
```diff
@@ -0,0 +1,144 @@
+
+type deleteResponse struct {
+	Subject  string `json:"subject" yaml:"subject"`
+	Versions []int  `json:"versions,omitempty" yaml:"versions,omitempty"`
```
maybe versions_deleted, to match the column header below?

twmb commented on src/go/rpk/pkg/cli/registry/schema/schema.go:
```diff
@@ -0,0 +1,103 @@
+			ID:      s.ID,
+			Type:    s.Type.String(),
+		}
+		if isText {
+			tw.PrintStructFields(r)
```
It might be a bit of duplication but can all of the output printing come at the end, and right here we just build rows?

r-vasquez replied:
This is on purpose:

If is **text** you just want to print the row, if it's json/yaml you'll want to have the full slice and then `f.Format()`

In my first pass, I had a first for loop building the rows, and at the end, another for loop for printing. But this isn't optimal for users that have a lot of schemas since we'll have to iterate through the array twice.

twmb replied:
The time delay of iterating these rows twice is incomparably small in comparison to fetching the array and printing it, so much so that IMO we should structure the code around readability.

twmb commented on src/go/rpk/pkg/cli/registry/schema/schema.go:
```diff
@@ -0,0 +1,103 @@
+	}
+	// There are commands where a table for text format is fine, but a json
+	// list does not make sense (e.g: schema create)
+	if single && len(rows) == 1 {
+		if _, _, s, err := f.Format(rows[0]); !isText {
+			if err != nil {
+				return fmt.Errorf("unable to print in the required format %q: %v", f.Kind, err)
+			}
+			fmt.Println(s)
+			return nil
+		}
+	}
+	if _, _, s, err := f.Format(rows); !isText {
+		if err != nil {
+			return fmt.Errorf("unable to print in the required format %q: %v", f.Kind, err)
+		}
+		fmt.Println(s)
+		return nil
```
I think these two blocks can be changed to something like
```

print := any(rows)
if single {
  print = rows[0]
}
if isText, _, s, err := f.Format(print); !isText {
  ...
  return
}
tw := out.NewTable(...)
defer tw.Flush()
for _, row := range rows {
  ...
}
```
and then get rid of the isText / a bunch of the logic at the top of the func?

twmb commented on src/go/rpk/pkg/cli/registry/schema/schema.go:
```diff
@@ -0,0 +1,103 @@
+type subjectSchema struct {
+	Subject string `json:"subject" yaml:"subject"`
+	Version int    `json:"version" yaml:"version"`
+	ID      int    `json:"ID" yaml:"ID"`
```
Is uppercase ID normal in json? I feel like it's normally lowercase -- I'd err on lowercase here.

gene-redpanda replied:
Second this, json is conventionally lowercase

twmb commented on src/go/rpk/pkg/cli/registry/schema/list.go:
```diff
@@ -0,0 +1,125 @@
+					}
+				}
+			}
+			if _, _, s, err := f.Format(response); !isText {
+				out.MaybeDie(err, "unable to print in the required format %q: %v", f.Kind, err)
+				out.Exit(s)
+			}
```
Same comment I have in schema.go -- can the struct building logic be above the format / print logic? It keeps the logic a bit easier to follow.

twmb commented on src/go/rpk/pkg/cli/registry/schema/list.go:
```diff
@@ -0,0 +1,125 @@
+type schemaResponse struct {
+	Subject string `json:"subject" yaml:"subject"`
+	Version int    `json:"version,omitempty" yaml:"version,omitempty"`
+	ID      int    `json:"ID,omitempty" yaml:"ID,omitempty"`
```
Same lowercase id comment as in schema.go

twmb commented on src/go/rpk/pkg/cli/registry/schema/get.go:
```diff
@@ -0,0 +1,129 @@
+		},
+	}
+
+	cmd.Flags().StringVar(&sversion, "version", "", "Version to lookup, or -1 or latest for the latest version, requires subject")
```
`, requires subject` => `; requires subject`, same comment on the --schema flag two lines below

twmb commented on src/go/rpk/pkg/cli/registry/schema/get.go:
```diff
@@ -0,0 +1,129 @@
+	cmd.Flags().StringVar(&sversion, "version", "", "Version to lookup, or -1 or latest for the latest version, requires subject")
+	cmd.Flags().IntVar(&id, "id", 0, "ID to lookup schemas usages of, subject optional")
+	cmd.Flags().StringVar(&schemaFile, "schema", "", "Schema file to check existence of, must be .avro or .proto, requires subject")
+	cmd.Flags().BoolVar(&deleted, "deleted", false, "If true, allow getting deleted schemas")
```
`allow getting` => `also return`

twmb commented on src/go/rpk/pkg/cli/registry/schema/get.go:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	cmd.Flags().StringVar(&sversion, "version", "", "Version to lookup, or -1 or latest for the latest version, requires subject")
+	cmd.Flags().IntVar(&id, "id", 0, "ID to lookup schemas usages of, subject optional")
```
`, subject optional` => `; subject required`

Also given how this is phrased `subject optional`, I'd change the `requires subject` just above and below to `subject required`

r-vasquez replied:
> , subject optional => ; subject required
But you don't need a subject when looking by ID

twmb replied:
Sorry I typo'd while thinking about the second sentence I was about to write
`, subject optional` => `; subject optional`

twmb commented on src/go/rpk/pkg/cli/registry/schema/get.go:
```diff
@@ -0,0 +1,129 @@
+					Type:   t,
+				})
+				out.MaybeDieErr(err)
+				err = printSubjectSchemaTable(f, false, s)
+				out.MaybeDieErr(err)
```
Can these lines, printing and maybe-err-exit, be moved outside of the switch in all cases?
And then save `ss` to a variable in each case?

Basically: I want to ensure what we only call printSubjectSchemaTable once, with `f, false, ss...`, because this will help ensure if/when this function grows that we _always_ use `false` for the second argument. It helps reduce the risk that we in the future think it is ok to use `true`, which will cause a panic since the type in the help is a slice.

gene-redpanda commented on src/go/rpk/pkg/cli/registry/compatibility_level.go:
```diff
@@ -0,0 +1,154 @@
+			}
+			results := cl.CompatibilityLevel(context.Background(), subjects...)
+
+			err = printCompatibilityResult(results, f)
+			out.MaybeDieErr(err)
```
```suggestion
			out.MaybeDieErr(printCompatibilityResult(results, f))
```

Any reason to reassign err here rather than daisy chain?

r-vasquez replied:
Not really, just to keep things clean and separated in the `Run` function of every command. A force of habit I'll say :sweat_smile:

twmb replied:
fwiw I like the pattern aspect of these last lines

gene-redpanda replied:
Alrighty then 😂

gene-redpanda commented on src/go/rpk/pkg/cli/registry/subject.go:
```diff
@@ -0,0 +1,144 @@
+	var permanent bool
+	cmd := &cobra.Command{
+		Use:   "delete [SUBJECT...]",
+		Short: "Soft or hard delete subjects",
```
```suggestion
		Short: "Soft or hard deletion of subjects",
```

gene-redpanda commented on src/go/rpk/pkg/cli/registry/subject.go:
```diff
@@ -0,0 +1,144 @@
+			}
+		},
+	}
+	cmd.Flags().BoolVar(&permanent, "permanent", false, "Perform a hard (permanent) delete of the subject; requires a soft-delete first")
+	return cmd
```
There's not a force option to get it done in one?

r-vasquez replied:
Is this per the SR specification @twmb ?

twmb replied:
Schema's _must_ be soft deleted before they can be hard deleted. We could potentially do two deletes inside the function itself, though, rather than requiring an end user to do the same command twice (the second time with an extra flag)

gene-redpanda replied:
That's pretty much all I'd be looking for from an admin perspective

r-vasquez replied:
So `--force-permanent`? And we issue the 2 requests and indicate in the error message if we got an error in the soft or permanent deletion?

gene-redpanda replied:
Yeah pretty much

gene-redpanda replied:
Also fine from my perspective

r-vasquez replied:
Added `--force` flag to both `rpk registry subject delete` and `rpk registry schema delete`. Thanks!

gene-redpanda reviewed (dismissed):
Comments aside LGTM!

gene-redpanda commented on src/go/rpk/pkg/config/params.go:
```diff
@@ -1301,6 +1395,23 @@ func (c *Config) fixSchemePorts() error {
+		if err != nil {
+			return fmt.Errorf("unable to fix schema registry address %v: %w", a, err)
+		}
+		switch scheme {
+		case "":
+			if port == "" {
+				port = strconv.Itoa(DefaultSchemaRegPort)
+			}
+			p.SR.Addresses[i] = net.JoinHostPort(host, port)
+		case "http", "https":
+			continue // keep whatever port exists; empty ports will default to 80 or 443
+		default:
+			return fmt.Errorf("unable to fix schema registry address %v: unsupported scheme %q", a, scheme)
+		}
```
I'm surprised this isn't a function somewhere

r-vasquez replied:
Yeah, I was talking with Travis and we have something ~similar for the Admin API but it follows a different logic since we also use the console URL as a redpanda admin.

There is room for refactoring but it will involve changing more packages and testing the Admin API functions, so I think we can leave it for another PR since this is already too big

r-vasquez commented:
Last force pushes addressed all of your comments. I'll create new commits for the features requested:

- `--force` flag to `subject delete` and `schema delete` to override the soft-deletion requirement.
- New command to support `referenced-by`

twmb commented on src/go/rpk/pkg/cli/registry/schema/check_compatibility.go:
```diff
@@ -0,0 +1,86 @@
+		},
+	}
+
+	cmd.Flags().StringVar(&schemaFile, "schema", "", "Schema filepath to check, must be .avro or .proto")
```
--type flag on this command as well?

twmb commented on src/go/rpk/pkg/cli/registry/schema/delete.go:
```diff
@@ -0,0 +1,73 @@
+
+	cmd.Flags().StringVar(&sversion, "schema-version", "", "Schema version to delete (latest, 0, 1...)")
+	cmd.Flags().BoolVar(&isPermanent, "permanent", false, "Perform a hard (permanent) delete of the schema; requires a soft-delete first")
+	cmd.Flags().BoolVar(&force, "force", false, "Force a hard (permanent) delete of the schema; It performs a soft-delete first")
```
nit: `; It performs a soft-delete first` => `; performs a soft-delete first`
(I think "It" can be dropped?)

twmb replied:
Actually now that I see both flags -- should we just always do hard deletion if the user specifies --permanent, and drop the --force flag?

twmb commented on src/go/rpk/pkg/cli/registry/schema/get.go:
```diff
@@ -0,0 +1,126 @@
+
+	cmd.Flags().StringVar(&sversion, "schema-version", "", "Schema version to lookup (latest, 0, 1...); subject required")
+	cmd.Flags().IntVar(&id, "id", 0, "ID to lookup schemas usages of; subject optional")
+	cmd.Flags().StringVar(&schemaFile, "schema", "", "Schema file to check existence of, must be .avro or .proto; subject required")
```
--type flag here?

twmb commented on src/go/rpk/pkg/schemaregistry/client.go:
```diff
@@ -0,0 +1,72 @@
+		return nil, errors.New("no schema registry hosts specified and rpk.yaml is configured to not use a default cluster")
+	}
+
+	urls := p.SR.Addresses
+	for i, u := range p.SR.Addresses {
+		scheme, _, err := net.ParseHostMaybeScheme(u)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse your schema registry address %q: %v", u, err)
+		}
+		switch scheme {
+		case "http", "https":
+			continue
+		case "":
+			if p.SR.TLS != nil {
+				urls[i] = "https://" + u
+			} else {
+				urls[i] = "http://" + u
+			}
+		default:
+			return nil, fmt.Errorf("unsupported scheme %q in the schema registry address %q", scheme, u)
+		}
```
I'm looking at this again and I think this should go in the params loading code, when initializing the virtual profile?
OK to live here for now but, maybe worth moving in follow up

twmb commented on src/go/rpk/pkg/cli/registry/subject.go:
```diff
@@ -0,0 +1,159 @@
+								err = fmt.Errorf("unable to perform hard-deletion: %w", err)
+							}
+						} else {
+							err = fmt.Errorf("unable to perform soft-deletion: %w", err)
```
`unable to perform initial soft-deletion that is required for hard-deletion: %v`
^^ I think we should expand the wording a little bit because if I were to see "unable to perform soft-deletion" when I'm trying to hard delete, I'd wonder what's up

twmb commented on src/go/rpk/pkg/cli/registry/schema/schema.go:
```diff
@@ -0,0 +1,126 @@
+	switch {
+	case strings.HasSuffix(schemaFile, ".avro") || strings.HasSuffix(schemaFile, ".avsc"):
+		return sr.TypeAvro, nil
+	case strings.HasSuffix(schemaFile, ".proto"):
```
proto || protobuf, to match the typeFromFlag logic?

twmb commented on src/go/rpk/pkg/cli/registry/schema/get.go:
```diff
@@ -0,0 +1,121 @@
+	cmd.Flags().StringVar(&sversion, "schema-version", "", "Schema version to lookup (latest, 0, 1...); subject required")
+	cmd.Flags().IntVar(&id, "id", 0, "ID to lookup schemas usages of; subject optional")
+	cmd.Flags().StringVar(&schemaFile, "schema", "", "Schema file to check existence of, must be .avro or .proto; subject required")
+	cmd.Flags().StringVar(&schemaType, "type", "", "Schema type of the file used to lookup, one of protobuf or avro; overrides schema file extension")
```
Can we add autocompletion to the --type flag, maybe as a common func in the schema.go file?

twmb commented on src/go/rpk/pkg/cli/registry/schema/delete.go:
```diff
@@ -0,0 +1,73 @@
+					err = cl.DeleteSchema(cmd.Context(), subject, version, sr.HardDelete)
+					out.MaybeDie(err, "unable to perform hard-deletion: %v", err)
+				}
+				out.MaybeDie(err, "unable to perform soft-deletion: %v", err)
```
Same wording comment here that is on the other error soft delete err message below

r-vasquez commented:
Added 2 new commits for better readability:

- Added `--type` autocompletion and improved some error message wording as suggested by Travis.
- Added the `--reference` flag to pass schema references when creating or checking the compatibility of a schema. This commit includes the relevant ducktape changes to test this part of the code. (Both unit and dt tests)

twmb commented on src/go/rpk/pkg/cli/registry/schema/create.go:
```diff
@@ -34,6 +37,22 @@ This uploads a schema to the registry, creating the schema if it does not
+
+You may pass the references using the --reference flag, which accepts either a
+comma separated list of <name>:<subject>:<version> or a path to a file. The file 
+must contain lines of name|subject|version separated by a tab or space, or the
```
`|` reads as if this must be formatted with pipe -- what do you think about "name, subject, and version"?

twmb commented on src/go/rpk/pkg/cli/registry/schema/create.go:
```diff
@@ -34,6 +37,22 @@ This uploads a schema to the registry, creating the schema if it does not
+  rpk registry schema create foo --schema /path/to/file --type avro
+
+Create a protobuf schema that references the schema in subject bar, version 1:
+  rpk registry schema create foo --schema /path/to/file.proto --references my_name:bar:1
```
Maybe `my_name:my_subject:1` rather than bar, to be clearer?

twmb commented on src/go/rpk/pkg/cli/registry/schema/check_compatibility.go:
```diff
@@ -26,7 +27,12 @@ type compatCheckResponse struct {
+	)
 	cmd := &cobra.Command{
 		Use:   "check-compatibility SUBJECT",
 		Short: "Check schema compatibility with existing schemas in the subject",
```
Long help text in future PR

r-vasquez commented:
Last [force push:](https://github.com/redpanda-data/redpanda/compare/8e1752c2a4dfc7dafc94b5483677549de6990b3a..0a00f2dbcc29599033f8def4d65654c63f3b83e0)

- Wording changes requested in the review.
